From 2858fea33c9719ba6dcf954e99b80c48bfddd955 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Wed, 26 Nov 2025 17:25:30 +0800
Subject: [PATCH 016/105] mtk: mt76: mt7996: only disallow getting global wcid
 in mt7996_rx_get_wcid

This commit refactors Felix'c change
(fb7e0501b8f920b22811f1900255a5646abf18b8)

The original commit causes the wcid index to be retrieved incorrectly
when using vif wcid for RX.

For example, after detecting beacon loss from the AP, STA will sends
a null func to AP and check if AP is still alive based on the TxS.
However, because the STA uses the vif's wcid to send the frame, the
wcid of the TxS cannot be correctly retrieved.

Now we only prevent global wcid to be retrieved in the RX path.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 mt7996/mac.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mt7996/mac.c b/mt7996/mac.c
index 2654830a9..bc482545d 100644
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -23,7 +23,7 @@ static struct mt76_wcid *mt7996_rx_get_wcid(struct mt7996_dev *dev,
 	int i;
 
 	wcid = mt76_wcid_ptr(dev, idx);
-	if (!wcid || !wcid->sta)
+	if (!wcid || wcid == &dev->mt76.global_wcid)
 		return NULL;
 
 	if (!mt7996_band_valid(dev, band_idx))
-- 
2.45.2

